src/control/tools/VerticalToolHandler.cpp

```diff
@@ -50,45 +42,106 @@ VerticalToolHandler::~VerticalToolHandler() {
+    cairo_restore(cr);
+
+    // if below, render elements translated so that startY is 0.
+    // if above, 0 is already the top of the page.
+    if (side > 0) {
+        cairo_translate(cr, 0, -this->startY);
+    }
+    DocumentView v;
+    v.drawSelection(cr, this);
+
+    cairo_destroy(cr);
+}
+
 void VerticalToolHandler::paint(cairo_t* cr, GdkRectangle* rect, double zoom) {
     GdkRGBA selectionColor = view->getSelectionColor();
 
     cairo_set_line_width(cr, 1);
 
     gdk_cairo_set_source_rgba(cr, &selectionColor);
 
-    double y = NAN;
-    double height = NAN;
-
-    if (this->startY < this->endY) {
-        y = this->startY;
-        height = this->endY - this->startY;
-    } else {
-        y = this->endY;
-        height = this->startY - this->endY;
-    }
+    const double y = std::min(this->startY, this->endY);
+    const double dy = this->endY - this->startY;
 
-    cairo_rectangle(cr, 0, y * zoom, this->page->getWidth() * zoom, height * zoom);
+    cairo_rectangle(cr, 0, y * zoom, this->page->getWidth() * zoom, std::abs(dy) * zoom);
 
     cairo_stroke_preserve(cr);
     auto applied = GdkRGBA{selectionColor.red, selectionColor.green, selectionColor.blue, 0.3};
     gdk_cairo_set_source_rgba(cr, &applied);
     cairo_fill(cr);
 
-    cairo_set_source_surface(cr, this->crBuffer, 0, this->endY * zoom);
+
+    const double elemY = (this->spacingSide > 0 ? this->endY : dy) * zoom;
+    cairo_set_source_surface(cr, this->crBuffer, 0, elemY);
     cairo_paint(cr);
+
+#ifdef DEBUG_SHOW_PAINT_BOUNDS
+    cairo_rectangle(cr, 0, elemY, cairo_image_surface_get_width(this->crBuffer),
+                    cairo_image_surface_get_height(this->crBuffer));
+    cairo_set_source_rgba(cr, 1.0, 0, 0, 0.3);
+    cairo_fill(cr);
+#endif
 }
 
 void VerticalToolHandler::currentPos(double x, double y) {
     double ySnapped = snappingHandler.snapVertically(y, false);
     if (this->endY == ySnapped) {
         return;
     }
-    double y1 = std::min(this->endY, ySnapped);
 
+    const double oldEnd = this->endY;
     this->endY = ySnapped;
 
-    this->view->repaintRect(0, y1, this->page->getWidth(), this->page->getHeight());
+    if (this->spacingSide > 0) {
+        this->view->rerenderRect(0, std::min(oldEnd, ySnapped), this->page->getWidth(), this->page->getHeight());
+    } else {
+        this->view->rerenderRect(0, 0, this->page->getWidth(), std::max(oldEnd, ySnapped));
+    }
+}
+
+bool VerticalToolHandler::onKeyPressEvent(GdkEventKey* event) {
+    if ((event->keyval == GDK_KEY_Control_L || event->keyval == GDK_KEY_Control_R) && this->spacingSide > 0) {
+        this->adoptElements(-1.0);
+        this->view->rerenderPage();
+        return true;
+    }
+    return false;
+}
+
+bool VerticalToolHandler::onKeyReleaseEvent(GdkEventKey* event) {
+    if ((event->keyval == GDK_KEY_Control_L || event->keyval == GDK_KEY_Control_R) && this->spacingSide < 0) {
+        this->adoptElements(1.0);
+        this->view->rerenderPage();
```

Comment:
Could we reduce this to `rerenderRect`? Rendering the full page should be a last resort.